inginious/frontend/plugins/upcoming_tasks/__init__.py

```diff
@@ -0,0 +1,146 @@
+        time_planner = "unlimited"
+        return self.page(time_planner)
+
+    def POST_AUTH(self):
+        """ Called when modifying time planner """
+        user_input = flask.request.form
+        time_planner = user_input.get("time_planner", default="unlimited")
+        return self.page(time_planner)
+
+    def time_planner_conversion(self, string_time_planner):
+        """ Used to convert the time_planner options into int value """
+        if string_time_planner in ["7", "14", "30"]:
+            return int(string_time_planner)
+        return 100000
+
+    def page(self, time_planner):
+        """ General main method called for GET and POST """
+        username = self.user_manager.session_username()
+        all_courses = self.course_factory.get_all_courses()
+        time_planner = self.time_planner_conversion(time_planner)
+
+        # Get the courses id
+        open_courses = {courseid: course for courseid, course in all_courses.items()
+                        if self.user_manager.course_is_open_to_user(course, username, False) and
+                        self.user_manager.course_is_user_registered(course, username)}
+        open_courses = OrderedDict(sorted(iter(open_courses.items()), key=lambda x: x[1].get_name(self.user_manager.session_language())))
+
+        # Get last submissions for left panel
+        last_submissions = self.submission_manager.get_user_last_submissions(5, {"courseid": {"$in": list(open_courses.keys())}})
+        except_free_last_submissions = []
+        for submission in last_submissions:
+            try:
+                submission["task"] = open_courses[submission['courseid']].get_task(submission['taskid'])
+                except_free_last_submissions.append(submission)
+            except:
+                pass
+
+        # Get the courses tasks, remove finished ones and courses that have no available unfinished tasks with upcoming deadline in range
+        tasks_data = {}
+        succeeded_courses = []
+        for courseid in open_courses:
+            course = open_courses[courseid]
+            outdated_tasks = []
+            tasks_data[courseid] = {}
+            tasks = course.get_tasks()
+            for taskid in tasks:
+                the_task = tasks[taskid]
+                if (not the_task.get_accessible_time().is_open()) or ((the_task.get_accessible_time().get_soft_end_date()) > (datetime.now()+timedelta(days=time_planner))):
+                    outdated_tasks += [taskid]
+            new_user_task_list = course.get_task_dispenser().get_user_task_list([username])[username]
+            tasks_data[courseid] = {taskid: {"succeeded": False, "grade": 0.0} for taskid in new_user_task_list}
+            user_tasks = self.database.user_tasks.find({"username": username, "courseid": course.get_id(), "taskid": {"$in": new_user_task_list}})
+            for user_task in user_tasks:
+                if not tasks_data[courseid][user_task["taskid"]]["succeeded"]:
```

Comment:
You don"t need to remove the succeeded ones anymore from the list ?

Reply:
I don't put the information in the tasks_data if it is succeeded.
Before I was putting information in tasks_data for all the tasks then removing the succeeded one, now I only add the non-succeeded ones :)

Reply:
Well, you are right, there is an issue here:
I am checking that tasks_data... is not succeed before even defining its content.
I should check based on user_data not on tasks_data

Reply:
I believe going back to the initial solution for this particular part:
```python
            for user_task in user_tasks:
                tasks_data[courseid][user_task["taskid"]]["succeeded"] = user_task["succeeded"]
                tasks_data[courseid][user_task["taskid"]]["grade"] = user_task["grade"]
                if tasks_data[courseid][user_task["taskid"]]["succeeded"]:
                    del tasks_data[courseid][user_task["taskid"]]
```python

might end up being the most simple solution.
To avoid setting grade and succeed the code would be:

```python
            for user_task in user_tasks:
                if not user_task["succeeded"]:
                    tasks_data[courseid][user_task["taskid"]]["succeeded"] = user_task["succeeded"]
                    tasks_data[courseid][user_task["taskid"]]["grade"] = user_task["grade"]
                else:
                    del tasks_data[courseid][user_task["taskid"]]
```

To me, it's as you prefer.

Reply:
well, you're right, the second option is probably better :)